### PR TITLE
Refund payments improvement

### PR DESCRIPTION
**Pull request by lukovskyj**
# GreenCityUBS PR

## Issue link

[#7302](https://github.com/ita-social-projects/GreenCity/issues/7302)

## Summary of change

Reworked refund logic:
- Removed endpoints for bonuses and money refunds.
- Integrated refund logic into PATCH /ubs/management/update-order-page-admin-info/{id}.
- Added additional checks for various refund scenarios.
- Updated refund logic, including three new fields in the table and connections with orders.
- Implemented user information display for refunded money and bonuses.
- Deleted unused classes and tests; added new tests.

## CHECK LIST
- [x]  СI passed
- [x]  Сode coverage >=95%
- [x]  PR is reviewed manually again (to make sure you have 100% ready code)
- [x]  All reviewers agreed to merge the PR
- [x]  I've checked new feature as logged in and logged out user if needed
- [x]  PR meets all conventions

### CODE REVIEW

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 1]
```diff
@@ -970,6 +1026,15 @@ private OrdersDataForUserDto getOrdersData(Order order) {
 
         Double amountBeforePayment = convertCoinsIntoBills(amountWithDiscountInCoins - paidAmountInCoins);
 
+        double refundedBonuses = order.getPayment().stream()
+            .filter(payment -> ENROLLMENT_TO_THE_BONUS_ACCOUNT_ENG.equals(payment.getReceiptLink()))
+            .map(payment -> payment.getAmount().doubleValue())
+            .reduce(0.0, Double::sum);
+
+        refundedBonuses = refundedBonuses == 0.0 ? 0.0 : refundedBonuses / (-100);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: can be simplified
` refundedBonuses /= -100;`

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 2]
```diff
@@ -970,6 +1026,15 @@ private OrdersDataForUserDto getOrdersData(Order order) {
...
+
+        Double refundedMoney = order.getRefund() == null ? 0.0 : -order.getRefund().getAmount().doubleValue() / 100;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: why do you divide by `-100` for `refundedBonuses` and by `100` for `refundedMoney`?
Please move `-100` to the constant and use it for both expressions

File: service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java [hunk 3]
```diff
@@ -1514,10 +1455,80 @@ public NotTakenOrderReasonDto getNotTakenOrderReason(Long orderId) {
             .build();
     }
 
-    @Override
-    public void saveOrderIdForRefund(Long orderId) {
-        Order order = orderRepository.findById(orderId)
-            .orElseThrow(() -> new NotFoundException(ORDER_WITH_CURRENT_ID_DOES_NOT_EXIST + orderId));
-        refundRepository.save(Refund.builder().orderId(order.getId()).build());
+    private void checkOverpayment(long overpayment) {
+        if (overpayment == 0) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:1459]: can `overpayment` be negative?
Should we add a check if `overpayment` is negative?

File: service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java [hunk 4]
```diff
@@ -1327,18 +1313,17 @@ public void updateOrderAdminPageInfoAndSaveReason(Long orderId, UpdateOrderPageA
...
-
-        if (order.getOrderStatus() == OrderStatus.BROUGHT_IT_HIMSELF
-            && nonNull(updateOrderPageDto.getGeneralOrderInfo())) {
-            updateOrderDetailStatus(order, updateOrderPageDto.getGeneralOrderInfo(), email);
-        } else {
-            try {
-                updateOrderPageFields(updateOrderPageDto, order, email);
-            } catch (Exception e) {
-                throw new BadRequestException(e.getMessage());
+        if (!processRefundForOrder(order, updateOrderPageDto, email)) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:1317]: I see that `processRefundForOrder` can throw an exception 
Should we add a catch block for this case?

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java (hunk 2 above)
Comment (lukovskyj) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: Currency conversion (kopiykas to hryvnias). `refundedBonuses` and `refundedMoney` are divided by `-100`.

File: service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java (hunk 4 above)
Comment (lukovskyj) [reply service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:1317]: Automatically handled by `@ExceptionHandler`

File: service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java (hunk 3 above)
Comment (lukovskyj) [reply service/src/main/java/greencity/service/ubs/UBSManagementServiceImpl.java:1459]: No, it cannot be negative. Here’s how the calculation is handled:

```
public static Long calculateOverpayment(Order order, Long sumToPayInCoins) {
        long paidAmountInCoins = calculatePaidAmount(order);

        long certificateSum = order.getCertificates().stream()
            .map(Certificate::getPoints)
            .reduce(0, Integer::sum);
        long bonusOverpaymentInCoins = 100L * (certificateSum + order.getPointsToUse()) - sumToPayInCoins;

        long overpaymentInCoins = paidAmountInCoins + bonusOverpaymentInCoins;

        return OrderStatus.CANCELED == order.getOrderStatus()
            ? paidAmountInCoins
            : Math.max(overpaymentInCoins, 0L);
    }
```

**Review by HelenSotnik — APPROVED** [review]

**Review by IliaWithHat — APPROVED** [review]

**Review by KizerovDmitriy — APPROVED** [review]

**Review by HelenSotnik — APPROVED** [review]

**Review by nazariusPr — APPROVED** [review]

**Pull request merged**